Fix lp:1669928 (Write/fsync amplification w/ duplicate GTIDs)

percona-ysorokin opened the pull request:
In some multi-source replications setups
'Xid_apply_log_event::do_apply_event()' method used to always update relay
log position regardless of whether the transaction being commited has already
been executed or not. In case of 'relay-log-info-repository=TABLE' this could
have had additional performance impact as updating 'slave_relay_log_info'
requires extra fsyncs. Moreover, depending on the number of channels receiving
duplicate (already executed) transactions, the situation could have become
proportionally worse.

Fixed by checking if the 'GTID_NEXT' was already logged and if so, omiting
relay log position update.

Added 'rpl.rpl_bug85141' MTR test case.

percona-ysorokin commented:
Jenkins build link
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/784/

laurynas-biveinis commented on sql/rpl_gtid.h:
```diff
@@ -3656,6 +3656,17 @@ int gtid_acquire_ownership_multiple(THD *thd);
+  @retval true   Transaction was already logged.
+  @retval false  Transaction must be executed.
+*/
+bool is_already_logged_transaction(const THD *thd);
```
const method

percona-ysorokin replied:
It was a static function outside of any class which I just made extern.

laurynas-biveinis commented on mysql-test/suite/rpl/t/rpl_bug85141.test:
```diff
@@ -0,0 +1,127 @@
+# Bug #85141 "Write/fsync amplification w/ duplicate GTIDs"
+#
+
+--source include/not_group_replication_plugin.inc
```
Add --source include/have_perfschema.inc

laurynas-biveinis commented on mysql-test/suite/rpl/t/rpl_bug85141.test:
```diff
@@ -0,0 +1,127 @@
+--echo #
+--let $rpl_connection_name = server_2
+--source include/rpl_connection.inc
+STOP SLAVE;
```
Async always looks suspicious in MTR - use include/stop_slave.inc include/start_slave.inc here and everywhere?

laurynas-biveinis commented on mysql-test/suite/rpl/t/rpl_bug85141.test:
```diff
@@ -0,0 +1,127 @@
+--source include/sync_slave_sql_with_master.inc
+--let $assert_text = number of records on server_3(C) after syncing with server_1(A) must be $number_of_records_on_master
+--let $assert_cond = COUNT(*) = $number_of_records_on_master FROM t1
+--source include/assert.inc
```
Consider include/rpl_diff.inc instead

laurynas-biveinis commented on mysql-test/suite/rpl/t/rpl_bug85141.test:
```diff
@@ -0,0 +1,127 @@
+--source include/rpl_end.inc
+
+
+#SELECT Channel_name, Relay_log_pos, Master_log_pos FROM mysql.slave_relay_log_info;
```
Leftovers?

percona-ysorokin commented:
5.6 is not affected as  Multi Source Replication was introduced in 5.7.2.

percona-ysorokin commented:
Relay log position is updated via abstract interface with 2 implementations (file repository and table repository). So, with 'relay-log-info-repository=FILE' we will also eliminate some unnecessary disk writes. I am just not sure how to measure number of relay log file updates in this case.

percona-ysorokin commented:
Addressed review comments: MTR test case now uses "include/rpl_diff.inc" to check data integrity on different servers.
Rebased on top of the most recent 5.7 to include fixed "include/rpl_diff.inc".
Updated Jenkins build link
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/809/
